### ISSUE TITLE
Splash screens on iOS

### ISSUE THREAD

**Issue by enchev** [NativeScript/ios-runtime#281]
_From @maknz on August 18, 2015 21:45_

In [this comment](https://github.com/NativeScript/nativescript-cli/issues/764#issuecomment-129833907), @ligaz mentioned it's possible to put launch/splash screens into `App_Resources/ios`. However, it doesn't seem to work off the bat (the default ones that shipped with {N} don't work either). Are there any guides available for how to achieve this?

_Copied from original issue: NativeScript/NativeScript#607_

**Comment by enchev** [comment]
_From @ligaz on August 19, 2015 4:26_

Hey @maknz,

The issue is that the default iOS project template includes a [LaunchScreen.xib](https://github.com/NativeScript/ios-runtime/blob/master/build/project-template/__PROJECT_NAME__/en.lproj/LaunchScreen.xib) file which takes a precedence over the defined splash screens resulting in white splash screen. We will have to double check do we still need this file to be in there by default. 

//cc @fealebenpae and @PanayotCankov

**Comment by jasssonpet** [comment]
@fealebenpae is this a duplicate of https://github.com/NativeScript/ios-runtime/issues/188

**Comment by fealebenpae** [comment]
IMO #188 is more related to https://github.com/NativeScript/nativescript-cli/issues/683.

I can only hazard a guess as to why we ended up shipping an empty LaunchScreen.xib in the first place - https://github.com/NativeScript/ios-runtime/commit/e3c40f2831f200aa57113b8fbbbda2519cb47eb8 claims it's related to the iPhone 6, but I'm afraid I can't remember any details regarding this particular decision.

[We're](https://github.com/NativeScript/ios-runtime/pull/267) [already](https://github.com/NativeScript/ios-runtime/pull/258) moving towards eradicating all vestiges of app icons and launch images from the NativeScriot iOS application template. IMO, it shouldn't have contained any of those in the first place.

As long as upstream consumers of the runtime (namely the CLI and AppBuilder) can verify that they can manage all requisite combinations of icons and images resolutions and pixel densities, I'll be more than happy to get rid of LaunchScreen.xib as well.

**Comment by ligaz** [comment]
I recall that there might be an issue with "zoomed" content on iPhone 6 Plus devices. I would vote to delete the the `LaunchScreen.xib` for now if everything works fine without it. @dtopuzov can you please check this when you have time?

**Comment by PanayotCankov** [comment]
The app was run in a different resolution when the launch screen is not provided. The cuteness showed two rows less as a result on iPhone6. The solution to add the LaunchScreen was quite rushed so I am :+1: for removing it. We should simply check the resolution is fine on iPhone6. Ping me if you need a hand to reproduce the problem.

**Comment by jlooper** [comment]
I had a lot of difficulty with shipping my iOS app due to launch screen problems. I ended up stripping all the launch screen images and not using an image asset library. Still, Xcode insisted that to add an app to the store, you need a 'tall screen' launch screen. So I just let Xcode create one itself and it added a black rectangular screen to the root with the required size and shape, then overwrote that file with one branded for my app:

![zgoauc-kwu-_b0zmef4zecvcue8zos9q9-ceuio8ckm](https://cloud.githubusercontent.com/assets/1450004/10137470/bc2a2c60-65c7-11e5-82bb-00852524987a.png)

**Comment by sfeather** [comment]
Jumping in a bit late here, but I'm for an either-or approach if possible.
A defined set of required image assets (checked at distribution build time for missing assets) or a xib if it can be made to work.

**Comment by fealebenpae** [comment]
We've removed the `LaunchScreen.xib` file - now all it takes to set a launch image is to drop a file with the proper dimensions and name (as per the [iOS HIG](https://developer.apple.com/library/ios/documentation/UserExperience/Conceptual/MobileHIG/LaunchImages.html)) in `App_Resources/iOS`.

**Comment by fealebenpae** [comment]
Reopened due to popular demand.

**Comment by kagha** [comment]
Yes please. Splash screen is an important feature.

I would like to simply be able to fill *App_Resources/iOS* and *App_Resources/Android* with the appropriate folder structure and file names.

If possible, I would like warnings during build if anything is missing or incorrectly structured.

**Comment by manijak** [comment]
This is a must!

**Comment by jlooper** [comment]
I'm battling this right now. Getting an app to the stores with the right splash screens is really hard now!

**Comment by manijak** [comment]
Cannot believe this is not prioritized... Please get it working in the next version or provide us a step-by-step tutorial on how to create proper splash screens for each platform. Thank you. 
@enchev @fealebenpae @ligaz

**Comment by benjaminpaul** [comment]
I am using 1.5.1 and I am still getting a black screen when the app starts which is actually stopping me from submitting it to the app store (they will decline it for this reason).

I have even tried editing the LaunchScreen.xib but even this gets ignored on start?

What is the correct workaround for the moment? Anything will do right now!

**Comment by fealebenpae** [comment]
@benjaminpaul I woud suggest removing your LaunchScreen.xib and switching to Asset Catalogs for your icons and launch screens in Xcode.

**Comment by wisonye** [comment]
@fealebenpae Hi, you said 
```javascript
I woud suggest removing your LaunchScreen.xib and switching to Asset Catalogs for your icons and launch screens in Xcode.
```

But actually how to do that, we're in emergency on this right now, could u please give us a instruction HOW TO DO THAT to get round the splash screen issue, Look forward to hear from, thank you SO MUCH !!! :)

**Comment by fealebenpae** [comment]
@wisonye I believe all you need to do is to open the Xcode project in `platforms/ios`, locate the `LaunchScreen.xib` file and delete it. Then you need to create the LaunchImage asset catalog as described [here](https://developer.apple.com/library/ios/recipes/xcode_help-image_catalog-1.0/chapters/StoringAppIconsandLaunchImagesinanAssetCatalog.html#//apple_ref/doc/uid/TP40013303-CH7-SW1). Once you create the asset catalog, you can open it in the Xcode editor and populate it with your images.

**Comment by wisonye** [comment]
@fealebenpae  Thanks, I did it already ,just watch few youtube video, anyway, thanks :)

**Comment by vchimev** [comment]
Write here to have in mind this crash in the splashboardd due to LaunchScreen.

```
splashboardd[16197]: *** Assertion failure in -[UINib initWithNibName:directory:bundle:], /SourceCache/UIKit_Sim/UIKit-3318.16.14/UINib.m:98
splashboardd[16197]: *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid parameter not satisfying: (name != nil) && ([name length] > 0)'
	*** First throw call stack:
	(
		0   CoreFoundation                      0x0000000105af3f35 __exceptionPreprocess + 165
		1   libobjc.A.dylib                     0x0000000105751bb7 objc_exception_throw + 45
		2   CoreFoundation                      0x0000000105af3d9a +[NSException raise:format:arguments:] + 106
		3   Foundation                          0x00000001052c25df -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 195
		4   UIKit                               0x00000001042139f2 -[UINib initWithNibName:directory:bundle:] + 164
		5   UIKit                               0x0000000104213aef +[UINib nibWithNibName:bundle:] + 61
		6   splashboardd                        0x00000001037e3d6d splashboardd + 7533
		7   libdispatch.dylib                   0x00000001063feaf4 _dispatch_client_callout + 8
		8   libdispatch.dylib                   0x00000001063ebe95 _dispatch_barrier_sync_f_slow_invoke + 51
		9   libdispatch.dylib                   0x00000001063feaf4 _dispatch_client_callout + 8
		10  libdispatch.dylib                   0x00000001063ea265 _dispatch_main_queue_callback_4CF + 490
		11  CoreFoundation                      0x0000000105a5bfe9 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
		12  CoreFoundation                      0x0000000105a1eeeb __CFRunLoopRun + 2043
		13  CoreFoundation                      0x0000000105a1e486 CFRunLoopRunSpecific + 470
		14  Foundation                          0x0000000105262b42 -[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 275
		15  Foundation                          0x00000001052eeb10 -[NSRunLoop(NSRunLoop) run] + 74
		16  splashboardd                        0x00000001037e429c splashboardd + 8860
		17  libdyld.dylib                       0x000000010642e145 start + 1
		18  ???                                 0x0000000000000001 0x0 + 1
	)
```

**Comment by raefa** [comment]
@manijak I also battled with this for a while ... and finally got it going. I am not sure if this is the best way to approach things ... but it seems to work. This may help: [Get Splash/Launch Screens working for iOS using NativeScript 1.6](http://blog.slay.com.au/post/139882861898/get-splashlaunch-screens-working-for-ios-using)

**Comment by PanayotCankov** [comment]
@raefa

We support now (1.6.0) the app `Info.plist` to be placed in `app/App_Resources/iOS/Info.plist` and it will be moved to the *platforms* upon build by the CLI. You can add images in `app/App_Resources/iOS/` and reference them from the `Info.plist` as launch images. The `platforms` folder is not intended to stay in source control, although in some cases there is no other choice. It would be really nice if you can manage to rework your article to use the `app/App_Resources/iOS/*` instead of the `platforms/`.

You can also take a look at the following article where the [launch images are mentioned](https://github.com/NativeScript/docs/blob/cankov/publishing/core-concepts/publishing-ios-apps.md#launch-screen-images). Feedback is welcome.

**Comment by raefa** [comment]
@PanayotCankov Sorry it took me a while to get back to you. Please excuse my ignorance but I don't see how I can put the images in the `app/App_Resources/iOS/`directory. Can you please explain? 

I note I had to create an image set and edit the "LaunchScreen.xib". As far as I am aware (and according to the documentation they have to be in the platforms directory.

> The NativeScript framework will provide a Launch Screen Storyboard in platforms/ios/<YourAppName>/en.lproj/LaunchScreen.xib, but does not yet provide a means to store it at app/App_Resources/iOS. If you want to edit it you can use the .xcodeproj generated in platforms/ios. You will have to add it in source control. CLI rebuilds may overwrite it, so you will have to watch out for automatic changes in it when you commit.
https://docs.nativescript.org/core-concepts/publishing-ios-apps.html#launch-screen-storyboard

**Comment by PanayotCankov** [comment]
You can take a look at the [nativescript-marketplace-demo](https://github.com/NativeScript/nativescript-marketplace-demo). Currently it uses images for launch screens. We will probably add a launch screen storyboard in future.

**Comment by raefa** [comment]
Ah yes ... I am aware you can just use images. I was actively trying to use the the LaunchScreen.xib. I initially had some issues with the images not showing so I pushed down the path of using the LaunchScreen.xib. Apple seems to be pushing that method anyway. I have added a note to my blog post stating that you can use images ... in case people stumble across it.

**Comment by PanayotCankov** [comment]
Ok, so for launch screen storyboards you can edit the `platform/ios/...` project. But the CLI may edit it too. Cleaning up the project and rebuilding it, may require some of the changes to be reapplied over and over.

**Comment by raefa** [comment]
Understood. :)

__________________________
*Raef Akehurst*

raefakehurst@gmail.com
+61 (0)402 884 406

**Comment by toddanglin** [comment]
@PanayotCankov @fealebenpae @valentinstoychev This seems like a rather "core" concept for shipping polished iOS apps. Any plans to get it prioritized in to an upcoming release?

**Comment by sitefinitysteve** [comment]
Just encountered this now as well... it's a headache that needs to be fixed

**Comment by bathejasumeet** [comment]
@all .
Guys.. there is a very easy way out as long as you are okay with static images used as splash .. tested it on nativescript 1.5.1 ..  and 1.7.1 
follow these steps on nativescript 1.5.1:
1) remove the launchscreen.xib file : delete it in xcode
2) In xcode select target under project explorer, select your project under targets
3)Select General on the top tab : Navigate to app Icons and Launch Images
4)Click use asset catalog for both app icons and launch images sources
5)Just migrate, no need to add any images in the assets manually in xcode
6)Leave the launch screen file section blank : this is very important : delete the text if present, else it will not work
7)run the app in emulator : you will see the default splash images under App_Resources/iOS being used, I believe replacing them should work

On nativescript 1.7.1, there is an extra step since contents of info.plist get replaced  : 
follow steps 1 to 6 above , after that, open info.plist under App_Resources / iOS , delete the key UILaunchStoryBoard along with the string LaunchScreen below it.


Tested it on emulators.

**Comment by vdende** [comment]
@all

I only followed the very last remark of @bathejasumeet. I just removed the following two files from `Info.plist`

>   `<key>UILaunchStoryboardName</key>`
>   `<string>LaunchScreen</string>`

Now it picks up the splash screen in my iPhone 6s 9.3 emulator. So I did nothing with xcode.
tns version 1.7.0

**Comment by bathejasumeet** [comment]
@vdende cool..

**Comment by bradmartin** [comment]
@vdende - that worked for me. I have hardly any iOS knowledge, thankfully I found your comment 👍

**Comment by vchittar** [comment]
@bathejasumeet @vdende Hey when I delete the lines on info.plist I am greeted by this error.

Error is: ENOENT: no such file or directory, stat '/Users/student/Desktop/Native_Script/Library/platforms/ios/Library/Resources/Info.plist

Any idea on how to fix this, it seems to be fixed by restoring the lines, but this doesnt give me my splash screen. Thanks!

**Comment by manijak** [comment]
@vchittar you aren't my any chance using livesync here? I've seen this type of error when modifying .plist files when in livesync. 

When modifying Info.plist you should always rebuild your app and restart livesync.

**Comment by vchittar** [comment]
@manijak Thanks, I did try that method before and it worked but whats curious is that the splashscreen still doesnt appear when I begin the app. I am sure there are the picture files in the appropriate folder.

**Comment by ignaciolarranaga** [comment]
Same for me, @vdende solution removing the key works. Thanks!

**Comment by vchimev** [comment]
@PanayotCankov, does your PR https://github.com/NativeScript/template-hello-world/commit/f7362743c717ce47ae873b02121a20968349d844 resolve this issue?

@NickIliev, is there any help article on this topic?

**Comment by PanayotCankov** [comment]
Yes,
This issue is duplicated multiple times.
https://github.com/NativeScript/nativescript-cli/issues/1868

**Comment by NickIliev** [comment]
@vchimev Yep - PR[ #411](https://github.com/NativeScript/docs/pull/411)

**Comment by felicks** [comment]
I'm still having this issue. Is there another solution for the current version of xcode and nativescript or does anybody had some trouble recently?

**Comment by NickIliev** [comment]
@felicks what is the issue you are dealing with? Follow the guidelines posted i[n this article](https://docs.nativescript.org/tooling/publishing/creating-launch-screens-ios#setting-launch-screen-and-app-icons) and rebuild the application (remove `platforms` folder before rebuilding) once all is done

**Comment by cremfert** [comment]
I'm having also an issue with this: my changes to the ```launchscreen.storyboard``` (even those which are applied to ```App_Resources/ios```) are overwritten each time I run ```tns run ios```. How can I make those permanent? Thanks!
